subsys/net/lib/downloader/src/downloader.c

```diff
@@ -0,0 +1,639 @@
+			return -EINVAL;
+		}
+		if (strncmp(hostname, dl->hostname, sizeof(hostname)) == 0) {
+			transport_connected = dl->transport;
+		}
+	}
+
+	/* Extract the hostname, without protocol or port */
+	err = dl_parse_url_host(uri, dl->hostname, sizeof(dl->hostname));
+	if (err) {
+		LOG_ERR("Failed to parse hostname, err %d", err);
+		k_mutex_unlock(&dl->mutex);
+		return -EINVAL;
+	}
+
+	/* Extract the filename, without protocol or port */
+	err = dl_parse_url_file(uri, dl->file, sizeof(dl->file));
+	if (err) {
+		LOG_ERR("Failed to parse filename, err %d, uri %s", err, uri);
+		k_mutex_unlock(&dl->mutex);
+		return -EINVAL;
+	}
+
+	dl->host_config = *host_config;
+	dl->file_size = 0;
+	dl->progress = from;
+	dl->buf_offset = 0;
+
+	dl->transport = NULL;
+	STRUCT_SECTION_FOREACH(dl_transport_entry, entry) {
+		if (entry->transport->proto_supported(dl, uri)) {
+			dl->transport = entry->transport;
+			break;
+		}
+	}
+
+	if (!dl->transport) {
+		if (strstr(uri, "://") == NULL) {
+			char *fallback = FALLBACK_HTTP;
+
+			if (host_config->sec_tag_list && host_config->sec_tag_count) {
+				fallback = FALLBACK_HTTPS;
+			}
+
+			LOG_WRN("Protocol not specified for %s, attempting %s", uri, fallback);
+			STRUCT_SECTION_FOREACH(dl_transport_entry, entry) {
+				if (entry->transport->proto_supported(dl, fallback)) {
+					dl->transport = entry->transport;
+					break;
+				}
+			}
+		}
+
+		if (!dl->transport) {
+			LOG_ERR("Protocol not found for %s", uri);
+			k_mutex_unlock(&dl->mutex);
+			return -EPROTONOSUPPORT;
+		}
+	};
+
+	if (is_state(dl, DOWNLOADER_CONNECTED)) {
+		if (dl->transport == transport_connected) {
+			state_set(dl, DOWNLOADER_CONNECTED, DOWNLOADER_DOWNLOADING);
+			goto out;
+		} else {
+			/* We are connected to the wrong host */
+			LOG_DBG("Closing connection to connect different host or protocol");
+			transport_connected->close(dl);
+			transport_connected->deinit(dl);
+		}
+	}
+
+	state_set(dl, (DOWNLOADER_IDLE | DOWNLOADER_CONNECTED), DOWNLOADER_CONNECTING);
+
+	memset(dl->transport_internal, 0, sizeof(dl->transport_internal));
+	err = transport_init(dl, &dl->host_config, uri);
+	if (err) {
+		state_set(dl, DOWNLOADER_CONNECTING, DOWNLOADER_IDLE);
+		k_mutex_unlock(&dl->mutex);
+		LOG_ERR("Failed to initialize transport, err %d", err);
+		return err;
+	}
+
+out:
+	k_mutex_unlock(&dl->mutex);
+
+	/* Let the thread run */
+	k_sem_give(&dl->event_sem);
+	return 0;
+}
+
+int downloader_cancel(struct downloader *const dl)
+{
+	if (dl == NULL ||
+	    is_state(dl, DOWNLOADER_IDLE) ||
+	    is_state(dl, DOWNLOADER_CONNECTED) ||
+	    is_state(dl, DOWNLOADER_DEINITIALIZED)) {
+		return -EPERM;
+	}
+
+	state_set(dl, STATE_ALLOW_ANY, DOWNLOADER_STOPPING);
+	return 0;
+}
+
+int downloader_get(struct downloader *dl,
+			const struct downloader_host_cfg *host_config,
+			const char *uri, size_t from)
+{
+	int rc;
+
+	if (dl == NULL || uri == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&dl->mutex, K_FOREVER);
```

Comment:
On one hand `host_config` is not checked because it's done in `downloader_start`, but then the same should apply to the mutex lock: there is no need to lock here because it is locked inside `downloader_start`. I think that this lock/unlock can be removed.